### ISSUE TITLE
Gradient Vanishing in Stacked Pure-Mamba Image Model (Seeking Advice)

### ISSUE THREAD

**Issue by LiulianC** [state-spaces/mamba#811]
Hello,
I’m currently encountering a *gradient vanishing* problem when training a **pure Mamba-based image processing model**. I would really appreciate your advice or any insights into possible causes and fixes.

### 🧩 Environment

* **PyTorch:** 2.6
* **Python:** 3.10
* **System:** Ubuntu
* **Mamba-ssm:** 2.2.2
* **Causal-conv1d:** 1.4.0 post 1
* Note: I cannot use the latest Mamba version because it reports an *illegal character* error: `_Nz3`.

### 🏗️ Model Architecture

I’m directly using the official classes:

https://github.com/state-spaces/mamba/blob/10b5d6358f27966f6a40e4bf0baa17a460688128/mamba_ssm/models/mixer_seq_simple.py#L118-L182

My network follows a typical *transformer-style backbone*:

```
image → patch embed → (Mamba1 ×10 + downsample) → (Mamba1 ×10 + downsample)
       → (Mamba1 ×10 + downsample) → (Mamba1 ×10) → decoder → B3HW
```

Each downsampling step doubles the number of channels and halves the spatial resolution.

I used the official initialization, residual connections, and feed-forward structure.
The initialization layers were **not overridden** by any global initialization.

### ⚠️ Problem: Gradient Vanishing

After training, the model output quality was poor, so I checked the gradients layer by layer.

I found that the gradients of the following parameters gradually vanished and propagated to other Mamba layers:

* `dt_proj.weight` and `dt_proj.bias`
* `mixer.fc1.weight` and `mixer.fc2.weight`
* `A_log`

Upon investigation:

* `dt_proj` is a linear projection layer for preparing parameters — it maps a low-rank temporal delta from `d_model/16` to `2*d_model`, a **32× expansion**, which might be quite large for a standard linear layer.

https://github.com/state-spaces/mamba/blob/10b5d6358f27966f6a40e4bf0baa17a460688128/mamba_ssm/modules/mamba_simple.py#L80

* `fc1` and `fc2` are the two linear layers inside the MLP, which also show gradient vanishing.

https://github.com/state-spaces/mamba/blob/10b5d6358f27966f6a40e4bf0baa17a460688128/mamba_ssm/modules/mlp.py#L24-L27

### 🙏 Request for Help

Resolving this gradient vanishing issue is very important for my current work.
I would be deeply grateful for any insights, debugging suggestions, or theoretical explanations the team could provide.

Thank you very much for your time and for the excellent open-source work on Mamba.

### 📜 Gradient Log (Partial)

[grad.log](https://github.com/user-attachments/files/23014210/grad.log)
